lib/ddtrace/contrib/grpc/datadog_client_interceptor.rb

```diff
@@ -0,0 +1,53 @@
+require 'grpc'
+require 'ddtrace/propagation/grpc_propagator'
+
+module Datadog
+  module Contrib
+    module GRPC
+      # The DatadogClientIntercptor implements the basic interface
+      # for gRPC middleware. When you declare Datadog usage of the gRPC
+      # module, the interceptor must be explicitly declared for usage.
+      class DatadogClientInterceptor < ::GRPC::ClientInterceptor
+        def request_response(request: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        def client_streamer(requests: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        def server_streamer(request: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        def bidi_streamer(requests: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        private
+
+        def ddtracer
+          @ddtracer ||= @options.fetch(:ddtracer, Datadog.tracer)
+        end
+
+        def trace_client(name, metadata = {})
+          metadata ||= {}
+          tracer_options = {
+            service: Datadog.configuration[:grpc][:service_name],
+            span_type: 'grpc'
+          }
+          span = ddtracer.trace(name, tracer_options)
+          Datadog::GRPCPropagator.inject!(span.context, metadata)
```

Comment:
This might be a silly question, but in regards to injecting the context info into the metadata...

If I'm understanding the code correctly, `GRPC` is supposed to provide a `metadata` Hash via `request_response(request: nil, call: nil, method: nil, metadata: nil)` options correct?  Is there a case where it doesn't actually provide this metadata Hash?

I ask here because above, `metadata ||= {}` will initialize the nil value to a Hash, but the `inject!` would end up modifying a local Hash variable that'd end up discarded anyways wouldn't it?

Reply:
That's a good question, and I'm not sure about the answer. I'll look it up. My suspicion is that the Hash is always supplied; I'm matching a method signature that I've seen elsewhere. I'll verify what's really needed here.